### ISSUE TITLE
reimplement unsaved warning

### ISSUE THREAD

**Issue by huss** [OpenEnergyDashboard/OED#1191]
### Describe the bug

Before RTK, OED used UnsavedWarning (searching for this will show where) to let the admin know when they might lose a change because it was unsaved. This was commented out to add in later after the basic RTK conversion was done.

### Expected behavior

These warning should be re-implemented and enabled in appropriate files.

OED currently does not, in general, use this in the new modal pages with cards. Thus, it is possible this will go away at some point. However, OED should decide if it should be used in some cases.

### Additional context

None

**Comment by gyordong** [comment]
Hi Steve, can @aduques and I work on this issue?

**Comment by huss** [comment]
> Hi Steve, can @aduques and I work on this issue?

@gyordong & @aduques You are welcome to work on this. I'm not sure how easy/hard it will be with the transition to using the Redux Toolkit. If it can be made to work we can discuss how it might be used with the modals. For example, would the warning pop up if the modal is closed (such as clicking off the modal) and there are unsaved changes? Please let me know if you need anything.

**Comment by aduques** [comment]
As it has been a couple of weeks since looking at this issue, I would like to verify the intended behavior of this feature. There is currently an unsaved warning that appears whenever the user tries to leave the Admin Settings page with unsaved changes. The pop-up can be closed out when clicking anywhere on the screen or by clicking “Cancel,” “Leave,” or “Save All.” This is currently implemented using React. This issue requires us to revert back and implement the unsaved warning feature using the corresponding Redux reducers, which have been commented out. Once the unsaved warning feature in the Admin Settings page is implemented with Redux, should similar pages, such as MapCalibrationComponent and MapsDetailComponent, have the same functionality and implementation?

At the moment, my partner and I will look to see how converting back to Redux can be done by looking through previous commits that have the Redux implemented and figure out how it can be carried over to the current version.

**Comment by huss** [comment]
Overall, I think what you stated is correct. It seems the src/client/app/components/UnsavedWarningComponent.tsx has had some changes so I'm not sure what might be needed to make it RTK compatible. src/client/app/types/redux/unsavedWarning.ts seems old but I'm unsure where that is used. src/client/app/components/HeaderButtonsComponent.tsx is involved and uses the old style in src/client/app/types/redux/actions.ts. I don't know if this helps but it is what I readily found.

As for maps, PR #1314 upgrades this to modals and RTK usage. It would be the starting point for any other changes. I am behind in getting that reviewed.

Finally, OED could add warnings to the modal pages for all the admin edits to avoid losing edits. It might be nice to do if it isn't too hard.

Please let me know if you need any other info or I can help.

**Comment by aduques** [comment]
After going through the commit history of the UnsavedWarningComponent.tsx, unsavedWarning.ts, and PreferencesComponent, I have come to an understanding of the following:

- The early implementation of the unsaved warning was implemented using both React and Redux.

- From commit e30fb1d to 72f111b, the implementation of the UnsavedWarningComponent changes from Redux to React. Commit e30fb1d indicated that there was an issue with the implementation of Redux, but from what I could understand, the commit itself addressed the issue.

- In commit 72f111b, In PreferencesComponent.tsx, the PreferencesProps were removed, and its contents were added to the adminSlice. This would include the method that initially handled the unsaved warning, but this specific method was not added to the adminSlice. Instead, the unsaved warning was implemented using React.

It appears that the initial decision to have the unsaved warning method excluded from the admin slice and be implemented using React was assigned in these issues:
https://github.com/OpenEnergyDashboard/OED/issues/726
https://github.com/OpenEnergyDashboard/OED/issues/886

I think I still need to look into this more, but I think that it is possible for the unsaved warning functionality to be implemented in the same manner as the other methods that have been moved from PreferencesProps to the adminSlice, as done in commit 72f111b. 

For this to be possible, I would need to start investigating how the current adminSlice functions to determine how the changes can be made and to also confirm if the adminSlice is implemented with Redux.

**Comment by ChrisMart21** [comment]
> * The early implementation of the unsaved warning was implemented using both React and Redux.
> 
> * From commit [e30fb1d](https://github.com/OpenEnergyDashboard/OED/commit/e30fb1df0f9972822e05f96cddd158750bc9fdfb) to [72f111b](https://github.com/OpenEnergyDashboard/OED/commit/72f111bb271848f9df94f6fb33f3bb2b63ce50a6), the implementation of the UnsavedWarningComponent changes from Redux to React. Commit [e30fb1d](https://github.com/OpenEnergyDashboard/OED/commit/e30fb1df0f9972822e05f96cddd158750bc9fdfb) indicated that there was an issue with the implementation of Redux, but from what I could understand, the commit itself addressed the issue.
> 
> * In commit [72f111b](https://github.com/OpenEnergyDashboard/OED/commit/72f111bb271848f9df94f6fb33f3bb2b63ce50a6), In PreferencesComponent.tsx, the PreferencesProps were removed, and its contents were added to the adminSlice. This would include the method that initially handled the unsaved warning, but this specific method was not added to the adminSlice. Instead, the unsaved warning was implemented using React.

One of the primary reasons for the omission/rework of this feature, was due to the pre-rtk implementation requiring storage of function(s)/callback(s) in the store. Functions are not plain/serializable data which is a cardinal rule that the Redux apps should abide by if possible. 
The original RTK migration attempted to squash many similar issues, but due to time constraints some features had to be left to a future commit.

[UnsavedWarningComponent.tsx](https://github.com/OpenEnergyDashboard/OED/blob/development/src/client/app/components/UnsavedWarningComponent.tsx) Is a solution I was experimenting with to utilize the some of the newer features from the react-router v6 bump that was also included in the inital rtk migration. IIRC I thought redux could be bypassed completely to fix/avoid the serializable issue, and simplify it a bit. It has some potential to be generic enough to be used anywhere a warning is needed, but it wasn't fleshed out completely due to time constraints. It should be a good enough starting point, or even inspiration for a better solution.

**Comment by huss** [comment]
Thanks to @ChrisMart21 for the insights. Those working on this should let us know if you need more information/help.

**Comment by aduques** [comment]
The reason as to why React and Redux are used together is because React simplifies the issue that Redux has when trying to store non-serializable data. Storing non-serializable data can cause problems later on, so using React as a temporary solution covers that. 

My plan for making the change from React to Redux is the following. I want to reimplement the React code that compares the local state of the preferences with the initial state by moving this functionality as an action in the adminSlice's reducer. However, doing so would inherently cause issues due to Redux's ability of not being able to support storing non-serializable data. To bypass this problem, I have done research and found out it is possible to make use of Redux's serializableCheck option to allow non-serializable data to be stored.

I'm aware that changing the serializableCheck option in Redux presents a lot of risks if not handled correctly. With my current understanding, I think that the current implementation of having Redux handle storing non-serializable data and React handling the routing is a much more safer option. I would like to clarify and ask again if serializableCheck is a potential solution that might work.

About serializableCheck from the Redux ToolKit documentation: https://redux-toolkit.js.org/usage/usage-guide#working-with-non-serializable-data 

My next steps, if plausible: 

- I want to take the React code from PreferencesComponent.tsx and make it as an action that will be added to the adminSlice's reducers in adminSlice.ts. 

- Once added to adminSlice, I want to update the store's serializableCheck to handle this new action in adminSlice's reducer. By doing so, it would address the storing non-serializable data issue that Redux has. 

- Once this implementation works for PreferencesComponent.tsx, I can implement the same unsaved warning feature to the relevant Component pages by updating the HTML code to take in the UnsavedWarningComponent.

- For the UnsavedWarningComponent.tsx, its implementation is done in React, but I think it would be possible to transition to Redux by following the same plan. However, I would need to look further into this.

Please let me know if there is anything that I'm misunderstanding or other factors I should be considering.

**Comment by huss** [comment]
Thanks to @aduques for continued efforts on this issue. I'm not an expert here so we can see if @ChrisMart21 has any thoughts.

I tried to find out if using ``serializableCheck`` would have a performance impact. I did not find info but did not easily see anything that indicates it would. During this search I found this discussion on addressing the non-serializable issue: https://github.com/reduxjs/redux-toolkit/discussions/1292. You have have already seen it but I wanted to pass it along.

**Comment by aduques** [comment]
Thank you for sharing the discussion as I have not read this one before. I have seen this suggestion (Put the logic related to the non-serialized object in a custom Redux middleware) often and it would be helpful for me to look more into this. Relying on a Redux middleware seems to be a way to handle non-serializable data.

According to the discussion, the recommended solution (Custom hook + global state) seemed to work best, but from my understanding, this solution would use both React and Redux. The custom hook itself would be implemented using Redux, but for the reducers in the adminSlice to call the function, React is still required.

Both of these solutions seem to require modifying serializableCheck in some way, so I will need to look more into it.

**Comment by ChrisMart21** [comment]
Much work was done to maintain serializable state, and to be able to safely enable the serializable dev check. I think it would be a step back, & in the wrong direction to put  unserializable data in the store. It's been a while since I've looked into this but typically there's a flaw in the approach if we need unserializable data in the store.

**Comment by huss** [comment]
@aduques I wanted to check the status of this work and see if you need anything.

**Comment by aduques** [comment]
Hello, I apologize for not working on this issue over the past few months, as my course load was more challenging than I anticipated. However, I can now devote time to the issue again.

If I am allowed to work on the issue again, I have some questions that will help me approach the issue.

A previous confusion that I had was that the goal of the issue was to make sure that the feature is “RTK compatible.” Can I get clarification on what is considered to be RTK compatible, as I previously thought that this meant Redux-only, but later learned that React + Redux might be a better choice.

Is the current implementation (in the [adminSlice.ts](http://adminslice.ts/) where the serializable data is handled in the adminSlice and the non-serializable data is handled with React calls (functions that handle the local changes in preferences), already considered to be RTK compatible? A point of confusion before was that I could not find a convenient way to allow Redux to handle this functionality.

In my notes, I noted that custom hooks are a possible solution for this issue, but I was unsure that it would follow a Redux-only implementation, as I have previously read that the solution involved React handling the function calls and Redux handling the state management.

To summarize, is working toward a React + Redux implementation suitable for the issue’s goal? If the current implementation that uses React + Redux is acceptable, should I work toward having all relevant pages that need unsaved warnings implemented in the same manner?

**Comment by huss** [comment]
Thanks to @aduques for getting back to me and for the questions.

> Hello, I apologize for not working on this issue over the past few months, as my course load was more challenging than I anticipated. However, I can now devote time to the issue again.
> 
> If I am allowed to work on the issue again, I have some questions that will help me approach the issue.

You are welcome to continue working on this.

> A previous confusion that I had was that the goal of the issue was to make sure that the feature is “RTK compatible.” Can I get clarification on what is considered to be RTK compatible, as I previously thought that this meant Redux-only, but later learned that React + Redux might be a better choice.

RTK is the Redux Toolkit that should work cleanly with React & Redux. We are trying to use it whenever possible. I'm not an expert but I think RTK has certain rules/desires that we want to follow. That would all OED to enable as many checks as possible to validate the code. Let me know if you need more information.

> Is the current implementation (in the [adminSlice.ts](http://adminslice.ts/) where the serializable data is handled in the adminSlice and the non-serializable data is handled with React calls (functions that handle the local changes in preferences), already considered to be RTK compatible? A point of confusion before was that I could not find a convenient way to allow Redux to handle this functionality.

I think it is. Using local state for page changes until stored (and then put into Global Redux state) is, I believe, a recommended way to do this. It is faster and means the app does not see the new state (in a global sense) until the page changes are stored. 

> In my notes, I noted that custom hooks are a possible solution for this issue, but I was unsure that it would follow a Redux-only implementation, as I have previously read that the solution involved React handling the function calls and Redux handling the state management.

I don't think using hooks is an issue. They are used a lot in OED. It is okay to use React/Redux. What we are trying to avoid, to support serializability, is having global Redux state that cannot be written out.

> To summarize, is working toward a React + Redux implementation suitable for the issue’s goal? If the current implementation that uses React + Redux is acceptable, should I work toward having all relevant pages that need unsaved warnings implemented in the same manner?

I think so as long as it is serializable. Again, I'm no expert. Please as for anything at any time.

@ChrisMart21 If you are available to look at/comment on this that would be nice since you know a lot more about this than I do.

**Comment by aduques** [comment]
Thank you for the helpful answers.

Another follow-up question: May I get clarification on where else the unsaved changes warning should appear (besides in the Admin Settings page)? I understand that a change that I should add is if the admin tries to log out while having any unsaved changes (in the admin setting page (or in any pages in general?)), the admin will be prompted with the unsaved warning popup and is prevented from being logged out. Currently, the admin can have unsaved changes in Admin Settings but can still log out, while the unsaved warning prompts whenever the user tries to leave the page.

It seems that the old unsavedWarning is added in the State interface ([state.ts](http://state.ts/)), however, aside from it being used in MapsDetailContainer.tsx, it does not seem to be relevant to UnsavedWarningComponent. With this in mind, am I correct to assume that [unsavedWarning.ts](http://unsavedwarning.ts/) can be abandoned for now, as it seems that [unsavedWarning.ts](http://unsavedwarning.ts/)’s UnsavedWarningState interface has been carried over and reimplemented as UnsavedWarningComponent.tsx’s UnsavedWarningProps’ interface.

If I want to achieve the goal of having the unsaved warning be prompted when an admin tries to log out with unsaved changes in the Admin Settings page, then should I update handleLogOut() in HeaderButtonsComponent.tsx to work with UnsavedWarningComponent.tsx rather than [unsavedWarning.ts](http://unsavedwarning.ts/)? I understand that to do this, I would need to update unsavedChangesState in HeadersButtonComponent.tsx accordingly.

**Comment by ChrisMart21** [comment]
state.ts was the older pre-rtk way of defining the shape of state. With rtk state is now inferred a bit differently. State.ts was kept for compatibility reasons, and should phased out. 
Unsaved warning.ts was used to define the shape of the older implementation of the unsavedChanges feature. It should also be phased out /omitted . 

The [HeaderButtonComponent](https://github.com/OpenEnergyDashboard/OED/blob/3799898d8e444f1daed8e1eba80f6d36db35289a/src/client/app/components/HeaderButtonsComponent.tsx#L77-L80) used to use the redux state that was tracking unsaved changes. The current implementation instead tracks the unsaved changes on a per page basis see [preferencesComponent](https://github.com/OpenEnergyDashboard/OED/blob/3799898d8e444f1daed8e1eba80f6d36db35289a/src/client/app/components/admin/PreferencesComponent.tsx#L33-L39) This decision was made due to time constrainsts and reduced redux complexity. I wasn't sure the direction of the feature and didn't invest too much time on figuring it out for intial migration. This same approach could be used on a per page basis for the other admin pages meterDetails, groupsDetails etc. 

As for the heaver buttons component, I am unsure with the current implementation how the headerbutton component would have access to the unsaved changes state, since it isn't in redux state, but on a per page basis (see preferencesComponent.tsx).

currently the [UnsavedWarningComponent](https://github.com/OpenEnergyDashboard/OED/blob/3799898d8e444f1daed8e1eba80f6d36db35289a/src/client/app/components/UnsavedWarningComponent.tsx#L27-L29) gets its unsaved changes state in by its parent using React-level state.

Somehow tracking the state in redux as before without the need to storing callbacks would be an approach. A simple boolean tracking unsaved changes which could be subsequently be selected by the component that needs it(unsavedWarning, headerbuttonscomponent etc...)
 Or even doing something more complex would work for example a slice dedicated to tracking updating local edits [localEditsSlice](https://github.com/ChrisMart21/OED/blob/e0f3af49acc0ee7040cdad04776cbedf56a36e28/src/client/app/redux/slices/localEditsSlice.ts#L88). 
[ComplimentaryCustomHook](https://github.com/ChrisMart21/OED/blob/e0f3af49acc0ee7040cdad04776cbedf56a36e28/src/client/app/redux/componentHooks.ts#L72-L135)

This was an experiment / playtest on what something like that could look like. 
Keep in mind that this was experimental and by no means the only approach. Neither was fully fleshed out and is open to the interpretation of the next developer who tackles it.

**Comment by huss** [comment]
I'm touching base to see what is going on for this issue. Please let me know any information or needed help. If it is not being actively done by the current assignee(s) then it is fine to let me know so I can plan what to do. Thanks.

**Comment by aduques** [comment]
I am currently working on this issue, but have had difficulties trying to model a solution similar to the custom hook approach that has been shared. Currently, I have opted for an implementation that reuses the current implementation in a custom hook so that it can be applied to all pages that would need the unsaved warning component. I am currently having issues with trying to use the custom hook that I've added to reimplement the 'const unsavedChangesState = false;' and the handleLogOut() in HeaderButtonsComponent.tsx so that it prevents admins from logging out and bypassing the unsaved warning. I am planning on pushing a pull request by the end of the weekend, so that my changes can be revised.

**Comment by huss** [comment]
Thanks to @aduques for the update. It is hard to give thoughts without seeing the code (and I'm not expert in this) so I'll wait until you push something. It is not a big rush so take the time you need. Many thanks for your continued work on this.

**Comment by huss** [comment]
PR #1504 moved this substantially forward. Issue #1517 contains the remaining work. Thus, I'm closing this issue.
